File: 2024/Golang/day3/part2.go
package day3

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func Part2() {
	file, err := os.Open("../Input/day3.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	sum := 0
	flag := true
	pattern := regexp.MustCompile(`(do\(\))|(don't\(\))|(mul\((\d+),(\d+)\))`)

	for scanner.Scan() {
		line := scanner.Text()
		matches := pattern.FindAllStringSubmatch(line, -1)

		for _, match := range matches {
			if match[1] != "" { 
				flag = true
			}
			if match[2] != "" { 
				flag = false
			}
			if match[3] != "" && flag { 
				firstNumber, _ := strconv.Atoi(match[4]) 
				secondNumber, _ := strconv.Atoi(match[5]) 
				sum += firstNumber * secondNumber
			}
		}
	}

	fmt.Println(sum)
}

File: 2024/Rust/day3/part2.rs
use crate::utils;
use regex::Regex;
use std::str::FromStr;

pub fn run_part2() {
    let re = Regex::new(r"(do\(\))|(don't\(\))|(mul\((\d+),(\d+)\))").unwrap();
    let mut sum = 0;
    let mut flag = true;

    for line in utils::read_lines("../Input/day3.txt") {
        for cap in re.captures_iter(&line) {
            if cap.get(1).is_some() {
                flag = true;
            }
            if cap.get(2).is_some() {
                flag = false;
            }
            if let (Some(first), Some(second)) = (cap.get(4), cap.get(5)) {
                if flag {
                    if let (Ok(first_number), Ok(second_number)) =
                        (i32::from_str(first.as_str()), i32::from_str(second.as_str()))
                    {
                        sum += first_number * second_number;
                    }
                }
            }
        }
    }

    println!("{}", sum);
}
